### PR TITLE
test: Fix race in TEST_recovery_scrub test

### PR DESCRIPTION
**Pull request by dzafman**
## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by neha-ojha — CHANGES_REQUESTED** [review]
@dzafman As discussed, we need to add 4 clear test cases noted in https://tracker.ceph.com/issues/46264#note-9, you ended up using a different tracker issue :)

**Comment by dzafman** [comment]
This pull request is on top of #37941 due to ef47a3e708da1b18865f2151e637ea12efdad8c1 which sets mon_allow_pool_size_one config when test wants pool size 1.

**Comment by dzafman** [comment]
@neha-ojha The only commit for this change is 43cac91ce22b0f867f42fd06c5a4fafb1d2a6c22.  I've implemented the first 2 tests.  The second test would be better if it really targeted a combination of recovery running on some OSDS and scrub running on non-intersecting OSDs.

**Comment by dzafman** [comment]
Testing of this pull request with #37941  passed:

https://pulpito.ceph.com/dzafman-2020-11-05_19:46:57-rados-wip-zafman-testing-48077-47930-distro-basic-smithi/

**Comment by dzafman** [comment]
> @dzafman As discussed, we need to add 4 clear test cases noted in https://tracker.ceph.com/issues/46264#note-9, you ended up using a different tracker issue :)

@neja-ojha
I corrected my use of the wrong tracker.  The list of 4 test cases in now in https://tracker.ceph.com/issues/47930 where they belong.  Do you want an implementation of all 4 tests?

**Comment by neha-ojha** [comment]
> > @dzafman As discussed, we need to add 4 clear test cases noted in https://tracker.ceph.com/issues/46264#note-9, you ended up using a different tracker issue :)
> 
> @neja-ojha
> I corrected my use of the wrong tracker. The list of 4 test cases in now in https://tracker.ceph.com/issues/47930 where they belong. Do you want an implementation of all 4 tests?

I am fine if we do it in phases(maybe, 1 and 2 first). These tests have been failing very regularly and I'd like to us to address it soon enough.

File: qa/standalone/scrub/osd-recovery-scrub.sh [hunk 1]
```diff
@@ -32,7 +32,107 @@ function run() {
...
+        return 1
+      fi
+      count=$(expr $count + 1)
+    done
+    set +o pipefail
+    ceph pg dump pgs
+
+    sleep 10
+    # Work around for http://tracker.ceph.com/issues/38195
+    kill_daemons $dir #|| return 1
```
Comment (neha-ojha) [inline qa/standalone/scrub/osd-recovery-scrub.sh:94]: why is this needed?

Comment (dzafman) [reply qa/standalone/scrub/osd-recovery-scrub.sh:94]: It isn't that important, but it freezes logs before greps.  The skipping of return 1, is due to intermittent failures of kill_daemon.

**Comment by dzafman** [comment]
Testing PASSED:

https://pulpito.ceph.com/dzafman-2020-11-10_23:41:09-rados-wip-zafman-testing-47930-2-distro-basic-smithi/

**Comment by dzafman** [comment]
@neha-ojha Ready to merge after final review

File: qa/standalone/scrub/osd-recovery-scrub.sh (hunk 1 above)
Comment (dzafman) [reply qa/standalone/scrub/osd-recovery-scrub.sh:94]: Both test cases are based on the original test as a starting point and the kill_daemons() was there.

**Comment by dzafman** [comment]
jenkins test api

File: qa/standalone/scrub/osd-recovery-scrub.sh [hunk 2]
```diff
@@ -42,11 +142,12 @@ function TEST_recovery_scrub() {
     OBJECTS=4
 
     setup $dir || return 1
-    run_mon $dir a --osd_pool_default_size=1 --mon_allow_pool_size_one=true || return 1
+    run_mon $dir a --osd_pool_default_size=1 --mon_allow_pool_size_one=true \
```
Comment (neha-ojha) [inline qa/standalone/scrub/osd-recovery-scrub.sh:145]: does this test still need 32 PGs?

Comment (dzafman) [reply qa/standalone/scrub/osd-recovery-scrub.sh:145]: I tried with 1 PG and the test failed.  Not sure why, but this works.

**Comment by dzafman** [comment]
Testing PASSED again

https://pulpito.ceph.com/dzafman-2020-11-13_09:37:50-rados-wip-zafman-testing-test-fixes-distro-basic-smithi/

File: qa/standalone/scrub/osd-recovery-scrub.sh (hunk 2 above)
Comment (neha-ojha) [reply qa/standalone/scrub/osd-recovery-scrub.sh:145]: If we can dig into the reason, we can make a follow-on fix for this. This is fine for now.

**Review by neha-ojha — APPROVED** [review]
